Solver: Increase ASP setup determinism

johnwparent opened the pull request:
Particular solver setup operations were performed in a manner allowing non determinism in the ordering of ASP rule generation. This PR ensures those operations are performed in an ordered fashion.

Resolving this allows for increased concretization cache hit rates

* Ensures sorting Spec objects is deterministic

Previously `spack.spec.Spec._cmp_iter` compared spec dependencies based on their spec hashes (among other attributes). This was insufficient for cases where abstract specs have dependencies that are conditional based on other dependencies introduced variance in two similar specs being compared, i.e.
in the case where a spec depends on pkg a when pkg b is version x and depends on package c when pkg b is version y.

This PR ensures a proper comparison is affected without unduly impacting spec comparison performance. 


Note: this PR is only part of the effort to enforce more determinism in ASP behavior in Spack, covering only the Spec related facets of that process, full determinism should not be expected until both this and #49471 land.

This PR in concert with #49471 produces the following benchmarks:
### Benchmarks

#### Develop Baseline:

##### `spack spec py-torch` timing from develop binary mirror w/ `--reuse`

```
real	        1m26.394s
user	1m15.934s
sys	        0m2.149s
```

##### Concretization cache hit rate for `py-torch`

0% (0/5)


##### Concretization cache hit rate for `zlib`

20% (1/5) - once 5 runs have been completed, hit rate is 100%

#### After these changes:

##### `spack spec py-torch` timing from develop binary mirror w/ `--reuse`
```
real	        1m22.092s
user	1m8.950s
sys	        0m2.623s
```

##### Concretization Cache hit rate for `py-torch`

100% (5/5)

##### Concretization Cache hit rate for `zlib`

100% (5/5)

Benchmarks performed on:
darwin-sonoma-m1

johnwparent commented:
@haampie @tgamblin @kwryankrattiger
We have determinism, this is now ready for review.
I'll be adding a test probably tomorrow.

johnwparent commented:
@spackbot fix style

haampie commented:
Removing `hash(...)` from `_cmp_iter` is good, cause it makes `<=` effectively random for abstract specs with dependencies.

However, you shouldn't `hash(...)` with `_cmp_iter` on dependencies though, cause it will make `s <= s` and `s == s` terribly slow:

```python
>>> import timeit
>>> s = spack.spec.Spec("./spec.json")
>>> t = spack.spec.Spec("./spec.json")
>>> timeit.timeit(lambda: s == t, number=1)
5.922160602000076
```

 that used to be

```python
>>> timeit.timeit(lambda: s == t, number=1)
0.013574219000020094
```

so 400x slower :)

The reason is of course it's following all possible paths from root to leaf, so you speed it up by just covering all edges.

But that will still slow down `s == s` for concrete specs, which is a case that's very important, because it happens whenever you do `s in set_of_specs` or `dict_of_specs[s]`.

Maybe an early return for the concrete specs makes sense.

tgamblin commented:
I'm a little confused b/c I do not see the slowdown Harmen does:

```python
#!/usr/bin/env spack-python

import timeit
from spack.spec import Spec

a = Spec("./hdf5.json")
b = Spec("./hdf5.json")

print(timeit.timeit(lambda: a == b, number=1))
```

with `hash()` -> `dep._cmp_iter`:
```console
# with change
> ~/time.py
0.015351500012911856
> ~/time.py
0.015207499964162707
> ~/time.py
0.01518858305644244

# with change
> ~/time.py
0.015128165949136019
> ~/time.py
0.015055165975354612
> ~/time.py
0.01517137500923127
```

Not much difference.

> Maybe an early return for the concrete specs makes sense, I didn't entirely think that through.

We already return the DAG hash (which requires zero additional computation) early for concrete specs, so I'm not sure what @haampie is seeing.

If there is an issue here, one idea that is admittedly kind of a band-aid would be to only do `hash()` for level 2 and lower. No abstract spec made outside of Spack's test suit is going to have dependencies lower than level 2, as they all come in attached at level 1. I don't think this is needed, though, as any spec with more levels is going to be concrete, unless I am missing something.

haampie commented:
> We already return the DAG hash (which requires zero additional computation) early for concrete specs, so I'm not sure what @haampie is seeing.

It's yielded early as a property to compare, but early return would only happen when they differ. For it to be slow you need a spec that has many possible paths from root to leaf, for example `py-black`.

johnwparent commented:
@spackbot fix style

johnwparent commented:
@tgamblin @haampie please see new approach to `_cmp_iter`

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -3770,9 +3770,6 @@ def eq_node(self, other):
     def _cmp_iter(self):
         """Lazily yield components of self for comparison."""
 
-        for item in self._cmp_node():
```
the root should go first, otherwise the order on concrete specs is by dag hash, which is equivalent to random.

johnwparent replied:
Sounds good, restoring this!

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -3790,13 +3787,13 @@ def _cmp_iter(self):
+
+            def yield_edge():
+                yield edge.spec._cmp_node
+                yield edge.depflag
```
edges have more properties than `depflag`.

potentially cleaner to `yield from spack.traverse_edges(...)` and make edges comparable -- but maybe first read https://github.com/spack/spack/pull/49391#pullrequestreview-2681859016.

johnwparent replied:
Makes sense, to a degree I was naively following the suggested diff so I can learn what's going on here. I think I have a better understanding now, not surprised I missed something. Will fix with that comment in mind.

haampie commented on lib/spack/spack/spec.py:
Can you split the changes to spec.py into a PR of its own? It would help with git bisect, cause it's easy to introduce bugs / breaking changes accidentally here.

johnwparent replied:
No problem, will do that ASAP, makes a lot of sense. 

Just want to point out that there will still be a great deal on non determinism in the asp setup output without the `_cmp_iter` changes as they resolve the primary source of non determinism. 

No issue with the request, just want to make it abundantly clear what to expect from the spun off "non determinism" PR without this as there were the concerns with the cache performance.

johnwparent replied:
Done, I left this PR containing the Spec changes and created a new one for the related but non spec determinism changes as I wanted to preserve the discussion and benchmarking you and Todd did.

You've been pinged and requested for review on all relevant PRs.

johnwparent replied:
Also thanks for making this comment on a line that does not effect the readability of the diff, idk how you commented on a line I didn't change, but appreciated all the same.

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -3790,13 +3787,13 @@ def _cmp_iter(self):
-                yield dep.spec.name
-                yield dep.depflag
-                yield hash(dep.spec)
+        for edge in self.traverse_edges(order="breadth", cover="edges"):
```
This is reasonable, but regresses the following:

```python
from spack.spec import Spec
import spack.deptypes as dt

a1, a2 = Spec("a"), Spec("a")
b1, b2 = Spec("b"), Spec("b")
c1, c2 = Spec("c"), Spec("c")
d1, d2 = Spec("d"), Spec("d")

c1.add_dependency_edge(d1, depflag=dt.ALL, virtuals=())
b1.add_dependency_edge(c1, depflag=dt.ALL, virtuals=())
a1.add_dependency_edge(c1, depflag=dt.ALL, virtuals=())
a1.add_dependency_edge(b1, depflag=dt.ALL, virtuals=())

c2.add_dependency_edge(d2, depflag=dt.ALL, virtuals=())
b2.add_dependency_edge(c2, depflag=dt.ALL, virtuals=())
a2.add_dependency_edge(c2.copy(deps=True), depflag=dt.ALL, virtuals=())
a2.add_dependency_edge(b2, depflag=dt.ALL, virtuals=())

assert a1 == a2
```

It's fixed by using the appropriate hash in `key=...` again, which effectively results in implementing `_cmp_iter` in terms of non-lazy `to_node_dict`.

haampie reviewed (changes requested):
Please split out the changes to `Spec._cmp_iter` into a PR of its own.

It's tedious to get it right, but getting it wrong is really bad not only because of non-determinism in `Spec.__lt__`, but rather because of false positives in `Spec.__eq__`. If you forget to yield just one property (like edge properties, missed in this PR) you break `x == y`. You can also not rely on tests catching these issues, cause apparently they don't exist.

After you've fixed the issue of `edge.virtuals`, you will have to go over all of

```
        yield self.name
        yield self.namespace
        yield self.versions
        yield self.variants
        yield self.compiler
        yield self.compiler_flags
        yield self.architecture
        yield self.abstract_hash
```

and see their respective `_cmp_iter` and check if those include all properties you also see in `to_node_dict`.

---

Taking a step back: the comments in the `_cmp_iter` method that two sources of truth are undesirable are on point. I think they are separate now because `_cmp_iter` is lazy and calling `to_node_dict` is expensive. But note that when you add `key=by_dag_hash` when imposing an order on the graph correctly, `to_node_dict` is called regardless to identify sub dags. I have to think a bit how to improve this situation.

johnwparent committed:
Spec: ensure set comparison is deterministic

Previously Spec equality comparison was performed
on the basis of root node attributes and dep hashes
but this did not appropriately address differences in
dependency constraints.
Now, comparison iterates through dependency nodes
to effect comparison for all nodes in the dag

johnwparent committed:
Refactor spec comparsion to traverse edges
